### DISCUSSION TITLE
How to enable plugins before jenv loads

### DISCUSSION THREAD

**Discussion by lujiajing1126** [zdharma-continuum/zinit#486] [Q&A]
I'm using jenv to manage all JVM versions on my computer.

```shell
#####################
# Java              #
#####################
zinit as'null' wait lucid atinit'export JENV_ROOT="$PWD"'\
    atclone'JENV_ROOT="$PWD" ./libexec/jenv init - > zjenv.zsh' \
    atpull"%atclone" src"zjenv.zsh" nocompile'!' sbin"bin/jenv" for \
        jenv/jenv
```

In order to export an env var `JAVA_HOME` to adapt Apache Maven, I have to enable `export` plugin in `jenv`. And this will generate a line in the `zjenv.zsh` file.

```shell
export PATH="/Users/megrez/.local/share/zinit/plugins/jenv---jenv/shims:${PATH}"
export JENV_SHELL=zsh
export JENV_LOADED=1
unset JAVA_HOME
unset JDK_HOME
source '/Users/megrez/.local/share/zinit/plugins/jenv---jenv/libexec/../completions/jenv.zsh'
jenv rehash 2>/dev/null
jenv refresh-plugins
# The following line must be added to load the plugin
source "/Users/megrez/.local/share/zinit/plugins/jenv---jenv/plugins/export/etc/jenv.d/init/export_jenv_hook.zsh"
jenv() {
  type typeset &> /dev/null && typeset command
  command="$1"
  if [ "$#" -gt 0 ]; then
    shift
  fi

  case "$command" in
  enable-plugin|rehash|shell|shell-options)
    eval `jenv "sh-$command" "$@"`;;
  *)
    command jenv "$command" "$@";;
  esac
}
```

I don't know in `zinit`, how can I accomplish this?

**Comment by vladdoster** [comment]
The following four lines will install `jenv` and set it up. It can replace your recipe and do what you want.

```zsh
zi ice as'command' pick'bin/jenv' id-as
zi light @jenv/jenv
zi ice lucid wait
zi snippet OMZP::jenv
```

Or using the `for` syntax

```
zi for \
    as'command' pick'bin/jenv' id-as light-mode \
  @jenv/jenv \
    lucid wait is-snippet \
  OMZP::jenv
```

In the image below:

- show that `jenv` is not installed.
- run the recipe above.
- show where it was installed and it works.

<img width="1312" alt="Screenshot 2023-03-23 at 21 58 09" src="https://user-images.githubusercontent.com/10052309/227413022-c6dd8fd9-ef14-4a9e-a1b0-b8988122fe46.png">

**Reply by lujiajing1126** [reply]
Thanks for your reply!

> Or using the `for` syntax
> 
> ```
> zi for \
>     as'command' pick'bin/jenv' id-as light-mode \
>   @jenv/jenv \
>     lucid wait is-snippet \
>   OMZP::jenv
> ```

But after switching to the code above, I'm not able to run any jenv commands. For example, running diagnosis with `jenv doctor`

![image](https://user-images.githubusercontent.com/2568208/227432757-74a27f3f-29bb-4e22-bf25-1592d9f79ea1.png)

**Reply by vladdoster** [reply]
@lujiajing1126, 

Am I understanding you correctly that my first recipe suggestion works as expected, but the second one using the `for` syntax does not work as expected? 



```
zi ice as'command' pick'bin/jenv' id-as
zi light @jenv/jenv
zi ice lucid wait
zi snippet OMZP::jenv
```

**Reply by lujiajing1126** [reply]
> @lujiajing1126,
> 
> Am I understanding you correctly that my first recipe suggestion works as expected, but the second one using the `for` syntax does not work as expected?
> 
> ```
> zi ice as'command' pick'bin/jenv' id-as
> zi light @jenv/jenv
> zi ice lucid wait
> zi snippet OMZP::jenv
> ```

It does not work as expected either. Running diagnosis `jenv doctor` will just output,

```
[OK]	No JAVA_HOME set
[ERROR]	Java binary in path is not in the jenv shims.
[ERROR]	Please check your path, or try using /path/to/java/home is not a valid path to java installation.
	PATH : /Users/megrez/.local/share/zinit/plugins/jenv/libexec:/Users/megrez/.local/share/zinit/plugins/pyenv---pyenv/shims:/Users/megrez/.nvm/versions/node/v16.19.1/bin:/Users/megrez/go/bin:/Users/megrez/.local/share/zinit/plugins/jenv/bin:/Users/megrez/.local/share/zinit/plugins/starship---starship:/Users/megrez/.local/share/zinit/polaris/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Users/megrez/.cargo/bin:/Users/megrez/Library/Application Support/JetBrains/Toolbox/scripts
[ERROR]	Jenv is not loaded in your zsh
[ERROR]	To fix : 	echo eval "$(jenv init -)" >> /Users/megrez/.zshrc
```

Probably I do not make my question clear. `jenv` is very similar to `pyenv`, so my snippet below works well in most cases,

```
zinit as'null' wait lucid atinit'export JENV_ROOT="$PWD"'\
    atclone'JENV_ROOT="$PWD" ./libexec/jenv init - > zjenv.zsh' \
    atpull"%atclone" src"zjenv.zsh" nocompile'!' sbin"bin/jenv" for \
        jenv/jenv
```

As you know, this will generate a so-called `zjenv.sh` and it will be sourced during initialization. But the actual issue is, I want to use plugins from `jenv`, e.g. `jenv enable-plugin export`. But after that, an extra `jenv init - > zjenv.zsh` must be called to generate a new source file. I am not sure how to do this in zinit. And in zsh, we can accomplish this by using the following codes,

```
echo 'export PATH="$HOME/.jenv/bin:$PATH"' >> ~/.zshrc
echo 'eval "$(jenv init -)"' >> ~/.zshrc
```

But so far in `zinit`, I have to call `jenv init -` once and use the output to override the original `zjenv.zsh`.
